bindings/py/tests/SDR_test.py

```diff
@@ -0,0 +1,318 @@
+        assert(A.getSparsity() == 0)
+        A.dense.fill(1)
+        A.setDenseInplace()
+        assert(A.getSparsity() == 1)
+
+    def testGetOverlap(self):
+        A = SDR((103,))
+        B = SDR((103,))
+        assert(A.getOverlap(B) == 0)
+
+        A.dense[:10] = 1
+        B.dense[:20] = 1
+        A.setDenseInplace()
+        B.setDenseInplace()
+        assert(A.getOverlap(B) == 10)
+
+        A.dense[:20] = 1
+        A.setDenseInplace()
+        assert(A.getOverlap(B) == 20)
+
+        A.dense[50:60] = 1
+        B.dense[0] = 0
+        A.setDenseInplace()
+        B.setDenseInplace()
+        assert(A.getOverlap(B) == 19)
+
+        # Test wrong dimensions
+        C = SDR((1,1,1,1, 103))
+        C.randomize( .5 )
+        try:
+            A.getOverlap(C)
+        except RuntimeError:
+            pass
+        else:
+            self.fail()
+
+    def testRandomizeEqNe(self):
+        A = SDR((103,))
+        B = SDR((103,))
+        A.randomize( .1 )
+        B.randomize( .1 )
+        assert( A != B )
+        A.randomize( .1, 0 )
+        B.randomize( .1, 0 )
+        assert( A != B )
+        A.randomize( .1, 42 )
+        B.randomize( .1, 42 )
+        assert( A == B )
+
+    def testAddNoise(self):
+        A = SDR((103,))
+        B = SDR((103,))
+        A.randomize( .1 )
+        B.setSDR( A )
+        A.addNoise( .5 )
+        assert( A.getOverlap(B) == 5 )
+
+        A.randomize( .3, 42 )
+        B.randomize( .3, 42 )
+        A.addNoise( .5 )
+        B.addNoise( .5 )
+        assert( A != B )
+
+        A.randomize( .3, 42 )
+        B.randomize( .3, 42 )
+        A.addNoise( .5, 42 )
+        B.addNoise( .5, 42 )
+        assert( A == B )
+
+    def testStr(self):
+        A = SDR((103,))
+        B = SDR((100, 100, 1))
+        A.dense[0] = 1
+        A.dense[9] = 1
+        A.dense[102] = 1
+        A.setDenseInplace()
+        assert(str(A) == "SDR( 103 ) 0, 9, 102")
+        A.zero()
+        assert(str(A) == "SDR( 103 )")
+        B.dense[0, 0, 0] = 1
+        B.dense[99, 99, 0] = 1
+        B.setDenseInplace()
+        assert(str(B) == "SDR( 100, 100, 1 ) 0, 9999")
+
+    @pytest.mark.skip(reason="Known issue: https://github.com/htm-community/nupic.cpp/issues/160")
+    def testPickle(self):
+        for sparsity in (0, .3, 1):
+            A = SDR((103,))
+            A.randomize( sparsity )
+            P = pickle.dumps( A )
+            B = pickle.loads( P )
+            assert( A == B )
+
+
+class SdrProxyTest(unittest.TestCase):
+    def testExampleUsage(self):
+        assert( issubclass(SDR_Proxy, SDR) )
+        # Convert SDR dimensions from (4 x 4) to (8 x 2)
+        A = SDR([ 4, 4 ])
+        B = SDR_Proxy( A, [8, 2])
+        A.sparse =  ([1, 1, 2], [0, 1, 2])
+        assert( (np.array(B.sparse) == ([2, 2, 5], [0, 1, 0]) ).all() )
+
+    def testLostSDR(self):
+        # You need to keep a reference to the SDR, since SDR class does not use smart pointers.
+        B = SDR_Proxy(SDR((1000,)))
```

Comment:
not for this PR, but this is tricky behavior. But probably having SDR_PRoxy manage collection of pointers would be as nasty.

Reply:
The purpose of this unit-test is the ensure that if/when the parent SDR is lost, the proxy doesn't crash & die.  The Proxy is notified when the parent is lost, and the proxy invalidates itself so this will never segfault, just throws an exception.

What's more worrying is:
```
python3
>>> A = SDR([1000])
>>> memory = A.dense
>>> del A
>>> memory -> SDR is gone, dense data is freed, but numpy array survives
```

Or more simply `SDR([1000]).dense` will allow access to memory that has been freed.